### ISSUE TITLE
Support for native half-float computation (float16/fp16)

### ISSUE THREAD

**Issue by benbarsdell** [tensorflow/tensorflow#4589]
https://github.com/tensorflow/tensorflow/issues/1300 added support for fp16 storage, but there is currently no support for native fp16 computation, which is available on some hardware such as Pascal GPUs.

In particular, the conv2d and matmul ops could take a new parameter along the lines of "compute_dtype", which would be plumbed through to CUDNN (convolution descriptor) and CUBLAS (Hgemm) in the backend, with the potential for up to a 2x speedup.

Related issues:
https://github.com/tensorflow/tensorflow/issues/1300
https://github.com/tensorflow/tensorflow/issues/4314
https://github.com/tensorflow/tensorflow/issues/851#issuecomment-230923665

**Comment by concretevitamin** [comment]
@benoitsteiner - Hello Benoit, since you've been looking at this, could you comment?

**Comment by agupta74** [comment]
Any update on the timeline for resolving this issue? Will it be fixed in the next TensorFlow software release (i.e v11)?

**Comment by Randl** [comment]
Still no updates on this one?

**Comment by dvader9** [comment]
Any update?

**Comment by wielandbrendel** [comment]
I'd also be interested in this feature! We just bought several P100 GPUs and are training on Imagenet a lot. Using FP16 would really speed things up.

**Comment by PhilJd** [comment]
I'm also very interested in this, as I'm currently developing a net optimized for inference speed. Would you accept pull requests for this? I guess it's mainly about what is described in this [ToDo](https://github.com/tensorflow/tensorflow/blob/904edee4456a61d50d5b1ffe9858a7772acc423e/tensorflow/stream_executor/cuda/cuda_dnn.cc#L1805) - find out which devices and CUDA versions support native float computation and then set the cudnn descriptor accordingly.

**Comment by Randl** [comment]
You need 5.3+ computability to use FP16

**Comment by PhilJd** [comment]
Started working on this ;)

**Comment by benoitsteiner** [comment]
We support native fp16 in most cases. However, we still compute convolutions and matrix multiplications using fp32 floats to avoid numerical stability issues.

In the case of the multiplications of a N by K matrix with a K by N matrix, we could use native fp16 provided that K is small. When K is large the reduced precision of fp16 introduces enough noise in the computation to cause issues such as reduced inference accuracy. Similarly, in the case of convolution, we could also use native fp16 provided that both the convolution window and the input depth are small. 

The trick here is going to be to figure out what thresholds are acceptable.

**Comment by PhilJd** [comment]
I played a bit with cuda fp16 ops - I'm not sure if its even useful to find such a threshold. The native fp16 op (i.e. hgemm) seems to be faster for larger square matrices only. [Here](https://github.com/hma02/cublasHgemm-P100) exists a benchmark of hgemm on a P100. As this is currently the only card where implementing these ops is useful (the  10xx cards support the fp16 ops but only with a major decrease in performance),  the fp16 ops are probably only interesting if future cuda/cudnn releases offer a speed up for computations on smaller matrices.

**Comment by benoitsteiner** [comment]
You might also want to take a look at the upcoming [Volta architecture](https://devblogs.nvidia.com/parallelforall/inside-volta/) that will supports mixed fp16/fp32 computation and should solve the problem.

**Comment by Randl** [comment]
Meanwhile caffe2 [adds 16 bit floating point](https://caffe2.ai/blog/2017/05/10/caffe2-adds-FP16-training-support.html) training support  in collaboration with NVIDIA.

**Comment by tfboyd** [comment]
This is being worked on.  Assigning to myself so I can update this as we go.

**Comment by slothkong** [comment]
Please keep us posted <3

**Comment by snarb** [comment]
Deffinetly we need this. My test on P100 shows great speed improvment in caffe fp16 for my model. Tensorflow now is still behinds for p100.

**Comment by tfboyd** [comment]
@reedwm Can you give some update and info on FP16?

**Comment by reedwm** [comment]
We're currently working on adding float16 ops to more TensorFlow ops, such as `tf.nn.fused_batch_norm`, as well as increasing the performance when using float16 ops. Although float16 ops use float32 internally for computations to avoid numerical precision issues, using float16 ops will result in increased performance over their float32 counterparts when run on P100 GPUs due to the reduced memory traffic.  Additionally, we plan on adding float16 support to [tf_cnn_benchmarks](https://github.com/tensorflow/benchmarks/tree/master/scripts/tf_cnn_benchmarks). 

I'll keep this bug updated as more progress is made.

**Comment by agupta74** [comment]
@reedwm Does it mean that TensorFlow does not plan to implement native FP16 (half-precision) FMA Cuda ops (opcode HFMA2) -- [https://devblogs.nvidia.com/parallelforall/mixed-precision-programming-cuda-8/](url) ? For inference there could be some applications which could have a significant speedup (theoretically 2x increase in FLOPS) and it is acceptable for these applications to have reduced precision?

**Comment by snarb** [comment]
News say that Intel will also add native  fp16 support for next Intel Xeon Phi processor (Knights Mill is commonly suggested) this year. There are few research that shows that in many cases fp16 is enouph for deep learning. Maybe not in all, but in many. Mb will be beter to allow the user make choice what presision is enouph? For example for my RL task I definitly know that fp16 presision  is  enough. And Cafee training acuracy for fp32 and fp16 is almost the same with much better speed for my task.

**Comment by reedwm** [comment]
@agupta74  Currently, float16 matmuls do compute in float16 if the environmental variable [TF_FP16_MATMUL_USE_FP32_COMPUTE](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/matmul_autotune.cc#L44) is set to "False". We plan on adding a similar environmental variable to convolutions. 

We are currently discussing whether to expose this as an argument, e.g. adding a `gpu_compute_type` argument to `tf.matmul`. Adding such an parameter forces us to keep it for a very long time due to our [backwards comparability guarantees](https://www.tensorflow.org/programmers_guide/version_semantics). This parameter would only apply to certain devices (GPUs with fp16 support) and will be unnecessary with Volta GPUs ([tensor cores](https://devblogs.nvidia.com/parallelforall/inside-volta/) will always be used on Volta devices) so we are still discussing whether the option is worth exposing without environmental variables.

@yaroslavvb Once we add float16 support to tf_cnn_benchmarks, we plan on reporting our training results.

**Comment by Sanghoon94** [comment]
I am trying to build a seq2seq model using fp16.
Using the same network architecture, fp32 model trains fine, but the model using fp16 suffers from CUDA_ERROR_OUT_OF_MEMORY.
Why is lower precision holds up larger memory? Is there any solution for this problem?

I am currently using python3, tf 1.0, cuda 8.0, cudnn 5.1

**Comment by tfboyd** [comment]
@Sanghoon94  Please open a different GitHub issue.  Without code you are asking for a guess that will not end up being very useful.  No guarantee someone will look at it but sharing the code allows for a better conversation.  You might also want to upgrade to TF 1.2.1 or TF 1.3 just to be more current as if anyone on the team tests the example you provide, they are going to use a much newer binary.  TF 1.0 is essentially 6 months old.

**Comment by rapatel0** [comment]
Hello @reedwm. Just wanted to know what the current status as of 1.3.1

**Comment by reedwm** [comment]
I don't think much has changed in 1.3.1. But on the master branch, the status is the following.

[tf_cnn_benchmarks](https://github.com/tensorflow/benchmarks/tree/master/scripts/tf_cnn_benchmarks) has support for fp16 by using the flag fp16. It currently does an unnecessary costly cast [here](https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/convnet_builder.py#L445), which I will remove once I verify doing so does not affect convergence. The cast is unnecessary since `tf.nn.fused_batch_norm` now supports fp16 on the GPU (and soon, on the CPU once #13388 is submitted).

By default, fp16 convolutions still do internal computations in fp32 (on pre-Volta hardware), but they can do fp16 internal computations by setting the newly added environmental variable [TF_FP16_CONV_USE_FP32_COMPUTE](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cuda_dnn.cc#L2136) to 0. As I stated before, fp16 matmuls also do compute in fp32, unless [TF_FP16_MATMUL_USE_FP32_COMPUTE](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/matmul_autotune.cc#L44) is set to 0.

**Comment by Randl** [comment]
@reedwm what about INT8? Any plans for its support?

**Comment by reedwm** [comment]
@petewarden can you comment on int8 support? 

Also, int8 support should probably be a separate issue.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tachim** [comment]
+1, any word on INT8?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yifeif** [comment]
@reedwm @tfboyd can we close this issue if fp16 support has been added? Thanks!
Feel free to open a separate issue for INT8.

**Comment by reedwm** [comment]
Closing since fp16 support has been added to many ops. If a particular op does not support fp16 but should, another issue can be filed. We also plan making mixed-precision training easier in the future by improving the API.

**Comment by UndeadBlow** [comment]
I converted model to pure DT_HALF, run Google Cloud with Nvidia Tesla T4 and model with DT_HALF is slower in about 8-9 times than DT_FLOAT.

Is this normal behaviour?
